File: backend/djadmin/assets/host_info.py
"""主机信息落库与按需拉取的公共逻辑。

原先主机静态信息由 Celery 周期任务通过 SSH(paramiko) 采集，现改为用户打开主机列表/详情页时
按需拉取：视图调用 refresh_host_info，经 gRPC 让 agent 执行 get_host_info。

Agent 启动时不主动采集主机快照。
"""
import logging
import uuid
import hashlib
import json

from django.db import transaction
from django.utils import timezone
from django.utils.dateparse import parse_datetime

from .models import Host, HostSystem, HostHardware, HostDisk, HostRuntime
from .grpc_transfer.client import AgentChannelClient, AgentGrpcTransferError

logger = logging.getLogger(__name__)


def _to_float(value):
    """安全地把上报值转为 float，非法或缺失返回 None。"""
    if value is None or value == '':
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _to_int(value):
    if value is None or value == '':
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _normalize_disks(disks):
    """规整磁盘列表，确保同一语义数据生成稳定哈希（与顺序无关）。"""
    normalized = []
    rows = disks if isinstance(disks, list) else []
    for item in rows:
        if not isinstance(item, dict):
            continue
        device = str(item.get('device') or '').strip()
        if not device:
            continue
        filesystem = str(item.get('filesystem') or '').strip() or None
        # squashfs 多为系统镜像只读挂载（如 snap），不纳入业务磁盘分区采集/展示。
        if str(filesystem or '').lower() == 'squashfs':
            continue
        normalized.append({
            'device': device,
            'mount_point': str(item.get('mount_point') or '').strip() or None,
            'filesystem': filesystem,
            'size_gb': _to_float(item.get('size_gb')),
            'used_gb': _to_float(item.get('used_gb')),
            'usage_percent': _to_float(item.get('usage_percent')),
        })
    normalized.sort(key=lambda item: (item.get('device') or '', item.get('mount_point') or ''))
    return normalized


def _build_static_fingerprint(result_data):
    """为静态资产字段构造稳定指纹，用于去重落库。"""
    source = result_data if isinstance(result_data, dict) else {}
    static_payload = {
        'os_type': str(source.get('os_type') or source.get('os') or '').strip() or None,
        'os_version': str(source.get('os_version') or '').strip() or None,
        'os_id': str(source.get('os_id') or '').strip().lower() or None,
        'os_id_like': str(source.get('os_id_like') or '').strip().lower() or None,
        'os_version_id': str(source.get('os_version_id') or '').strip() or None,
        'kernel_version': str(source.get('kernel_version') or '').strip() or None,
        'hostname': str(source.get('hostname') or '').strip() or None,
        'agent_version': str(source.get('agent_version') or '').strip() or None,
        'cpu_count': _to_int(source.get('cpu_count')),
        'cpu_model': str(source.get('cpu_model') or '').strip() or None,
        'memory_total_gb': _to_float(source.get('memory_total_gb')),
        'arch': str(source.get('arch') or '').strip() or None,
        'os_timezone': str(source.get('os_timezone') or '').strip() or None,
        'os_utc_offset': str(source.get('os_utc_offset') or '').strip() or None,
        'disks': _normalize_disks(source.get('disks')),
    }
    encoded = json.dumps(static_payload, ensure_ascii=False, sort_keys=True, separators=(',', ':'))
    return hashlib.sha256(encoded.encode('utf-8')).hexdigest()


def _to_datetime(value):
    parsed = parse_datetime(str(value or '').strip())
    if parsed is not None and timezone.is_naive(parsed):
        parsed = timezone.make_aware(parsed, timezone.utc)
    return parsed


def persist_host_info(host, status, result_data, error='', now=None):
    """将一次 get_host_info 的结果落库到 Host/HostSystem/HostHardware/HostDisk。

    result_data 由 dj-agent 的 get_host_info 返回，包含：
    - 静态资产：os_type/os_version/kernel_version/cpu_model/cpu_count/arch/memory_total_gb/disks；
    - 动态指标：cpu_usage_percent/memory/disk_io/os_uptime_* 等写入 HostRuntime。

    返回 True 表示成功落库静态资产；失败/无数据返回 False。
    """
    now = now or timezone.now()
    is_success = str(status or '').strip() == 'success'
    static_fingerprint = ''
    if is_success and isinstance(result_data, dict) and result_data:
        static_fingerprint = _build_static_fingerprint(result_data)
        result_data = dict(result_data)

    # Host 主表只保存采集状态；失败时不覆盖上次成功的 collect_time。
    update_fields = ['collect_status', 'collect_message', 'update_time']
    host.collect_status = 'success' if is_success else 'failed'
    host.collect_message = str(error or '')
    if is_success:
        host.collect_time = now
        update_fields.append('collect_time')
    host.save(update_fields=update_fields)

    if not is_success or not isinstance(result_data, dict) or not result_data:
        return False

    previous_runtime = HostRuntime.objects.filter(host=host).only('static_fingerprint').first()
    previous_fingerprint = str(getattr(previous_runtime, 'static_fingerprint', '') or '').strip()
    HostRuntime.objects.update_or_create(
        host=host,
        defaults={
            'cpu_usage_percent': _to_float(result_data.get('cpu_usage_percent')),
            'cpu_times': result_data.get('cpu_times') if isinstance(result_data.get('cpu_times'), dict) else {},
            'memory_usage_percent': _to_float(result_data.get('memory_usage_percent')),
            'memory': result_data.get('memory') if isinstance(result_data.get('memory'), dict) else {},
            'disk_io': result_data.get('disk_io') if isinstance(result_data.get('disk_io'), (list, dict)) else [],
            'os_uptime_seconds': _to_int(result_data.get('os_uptime_seconds')),
            'os_boot_time': _to_datetime(result_data.get('os_boot_time')),
            'metrics_sample_window_ms': _to_int(result_data.get('metrics_sample_window_ms')),
            'static_fingerprint': static_fingerprint,
            'collected_at': now,
        },
    )
    if previous_fingerprint and previous_fingerprint == static_fingerprint:
        # 静态资产未变化：保留最新动态快照即可，跳过系统/硬件/磁盘表重复写入。
        return True

    # os_type 优先取发行版名称（os_release NAME，如 "Ubuntu"），回退到 GOOS（"linux"）。
    os_type = str(result_data.get('os_type') or result_data.get('os') or '').strip() or None

    HostSystem.objects.update_or_create(
        host=host,
        defaults={
            'os_type': os_type,
            'os_version': str(result_data.get('os_version') or '').strip() or None,
            'os_id': str(result_data.get('os_id') or '').strip().lower() or None,
            'os_id_like': str(result_data.get('os_id_like') or '').strip().lower() or None,
            'os_version_id': str(result_data.get('os_version_id') or '').strip() or None,
            'kernel_version': str(result_data.get('kernel_version') or '').strip() or None,
            'hostname': str(result_data.get('hostname') or '').strip() or None,
            'agent_version': str(result_data.get('agent_version') or '').strip() or None,
            'timezone_name': str(result_data.get('os_timezone') or '').strip() or None,
            'utc_offset': str(result_data.get('os_utc_offset') or '').strip() or None,
            'collector_source': 'agent',  # 固定来源：dj-agent 本地采集
            'collected_at': now,
        },
    )

    disks = _normalize_disks(result_data.get('disks'))
    result_data['disks'] = disks
    disk_total_gb = None
    if disks:
        total = sum((_to_float(d.get('size_gb')) or 0) for d in disks if isinstance(d, dict))
        disk_total_gb = round(total, 1) if total > 0 else None

    HostHardware.objects.update_or_create(
        host=host,
        defaults={
            'cpu_cores': _to_int(result_data.get('cpu_count')),
            'cpu_model': str(result_data.get('cpu_model') or '').strip() or None,
            'memory_gb': _to_float(result_data.get('memory_total_gb')),
            'disk_total_gb': disk_total_gb,
            'architecture': str(result_data.get('arch') or '').strip() or None,
            'collected_at': now,
        },
    )

    # 磁盘表按最新挂载全量重建，以清除已卸载/删除的分区，避免残留脏数据。
    with transaction.atomic():
        HostDisk.objects.filter(host=host).delete()
        rows = []
        for d in disks:
            if not isinstance(d, dict):
                continue
            device = str(d.get('device') or '').strip()
            if not device:
                continue
            rows.append(HostDisk(
                host=host,
                device=device,
                mount_point=str(d.get('mount_point') or '').strip() or None,
                size_gb=_to_float(d.get('size_gb')),
                used_gb=_to_float(d.get('used_gb')),
                filesystem=str(d.get('filesystem') or '').strip() or None,
            ))
        if rows:
            HostDisk.objects.bulk_create(rows)

    return True


def refresh_host_info(host, timeout_seconds=15):
    """按需经 gRPC 让指定主机的 agent 执行 get_host_info，并落库。

    返回统一结构：{host_id, updated, skipped, error}
    - skipped=True：因缺少 agent_id 或 agent 离线而未发起拉取（非错误）；
    - updated=True：成功拉取并落库静态资产；
    - error：失败原因（skipped/失败时填充）。
    """
    result = {'host_id': host.id, 'updated': False, 'skipped': False, 'error': ''}

    agent_id = str(host.agent_id or '').strip()
    if not agent_id:
        result['skipped'] = True
        result['error'] = '主机未配置 agent_id，无法定位 agent'
        return result

    # 只对在线 agent 发起拉取，离线直接跳过（避免同步等待超时拖慢页面加载）。
    if not host.agent_online:
        result['skipped'] = True
        result['error'] = 'agent 离线'
        return result

    try:
        client = AgentChannelClient(agent_id)
        resp = client.execute_automation(
            job_id=uuid.uuid4().hex,
            params={},
            timeout_seconds=timeout_seconds,
            task_type='inventory',
            action='get_host_info',
        )
    except AgentGrpcTransferError as exc:
        # 通道级失败：记录为一次采集失败，但不抛出，交由调用方汇总。
        persist_host_info(host, 'failed', {}, error=str(exc))
        result['error'] = str(exc)
        return result
    except Exception as exc:  # noqa: BLE001 - 兜底防止单台异常影响批量
        logger.error('refresh_host_info unexpected error for %s: %s', agent_id, exc, exc_info=True)
        result['error'] = str(exc)
        return result

    status = str(resp.get('status') or '').strip()
    result_data = resp.get('result_data') or {}
    error_message = str(resp.get('error_message') or '')
    updated = persist_host_info(host, status, result_data, error=error_message)
    result['updated'] = updated
    if not updated and error_message:
        result['error'] = error_message
    return result

File: autoadmin/internal/assets/host_info.go
package assets

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"

	"autoadmin/internal/agent/pb"
	db "autoadmin/internal/platform/database/generated"
)

// hostInfoOutcome mirrors Django's refresh_host_info() return contract so the frontend's
// existing {result, host} handling keeps working unchanged.
type hostInfoOutcome struct {
	HostID  int64  `json:"host_id"`
	Updated bool   `json:"updated"`
	Skipped bool   `json:"skipped"`
	Error   string `json:"error"`
}

// refreshHostAgentInfo dispatches a synchronous get_host_info job to the host's agent and
// persists the result, mirroring the Django-era refresh_host_info + persist_host_info.
// Requires the caller to have already called applyAgentPresence(&host) for a fresh online check.
func (h *Handler) refreshHostAgentInfo(ctx context.Context, host Host) hostInfoOutcome {
	outcome := hostInfoOutcome{HostID: host.ID}

	instanceName := ""
	if host.InstanceName != nil {
		instanceName = strings.TrimSpace(*host.InstanceName)
	}
	if instanceName == "" {
		outcome.Skipped = true
		outcome.Error = "主机未配置实例名，无法定位 agent"
		return outcome
	}
	if !host.AgentOnline {
		outcome.Skipped = true
		outcome.Error = "agent 离线"
		return outcome
	}

	response, err := h.gateway.Execute(ctx, instanceName, &pb.AutomationExecuteRequest{
		JobId:          fmt.Sprintf("host-info-%d-%d", host.ID, time.Now().UnixNano()),
		Type:           "inventory",
		Action:         "get_host_info",
		ParamsJson:     "{}",
		TimeoutSeconds: 15,
	})
	if err != nil {
		if _, persistErr := h.service.persistHostInfo(ctx, host.ID, "failed", nil, err.Error()); persistErr != nil {
			outcome.Error = persistErr.Error()
			return outcome
		}
		outcome.Error = err.Error()
		return outcome
	}

	var resultData map[string]any
	if response.ResultDataJson != "" {
		_ = json.Unmarshal([]byte(response.ResultDataJson), &resultData)
	}
	updated, err := h.service.persistHostInfo(ctx, host.ID, response.Status, resultData, response.ErrorMessage)
	if err != nil {
		outcome.Error = err.Error()
		return outcome
	}
	outcome.Updated = updated
	if !updated && response.ErrorMessage != "" {
		outcome.Error = response.ErrorMessage
	}
	return outcome
}

// RefreshHostInfoByID 供其他域（如 monitor 的 exporter 目标安装）在缺少主机平台/架构信息时
// 主动补采一次资产信息并落库：同步向 agent 下发 get_host_info，成功后 hardware/system 等
// 平台字段即可用于选包。返回采集失败原因，供调用方决定是否继续或提示。
func (h *Handler) RefreshHostInfoByID(ctx context.Context, hostID int64) error {
	item, err := h.service.GetHost(ctx, hostID)
	if err != nil {
		return err
	}
	h.applyAgentPresence(&item)
	outcome := h.refreshHostAgentInfo(ctx, item)
	if outcome.Error != "" {
		return errors.New(outcome.Error)
	}
	return nil
}

// persistHostInfo writes one get_host_info result into assets_host/hostsystem/hosthardware/
// hostruntime/hostdisk, matching Django's persist_host_info() field-for-field.
func (s *Service) persistHostInfo(ctx context.Context, hostID int64, status string, resultData map[string]any, errorMessage string) (bool, error) {
	pool := s.repository.pool
	now := time.Now().UTC()
	isSuccess := status == "success"

	queries := db.New(pool)
	// 成功时写 collect_time；失败时传 NULL 保留上次采集时间（原实现是两条 UPDATE）。
	collectTime := sql.NullTime{}
	if isSuccess {
		collectTime = sql.NullTime{Time: now, Valid: true}
	}
	if err := queries.MarkHostCollected(ctx, db.MarkHostCollectedParams{
		CollectStatus: status, CollectMessage: errorMessage, CollectTime: collectTime,
		UpdateTime: now, ID: hostID,
	}); err != nil {
		return false, err
	}
	if !isSuccess || len(resultData) == 0 {
		return false, nil
	}

	disks := normalizeHostDisks(resultData["disks"])
	fingerprint := buildStaticFingerprint(resultData, disks)

	previousFingerprint, err := queries.GetHostRuntimeFingerprint(ctx, hostID)
	if err != nil && !errors.Is(err, sql.ErrNoRows) {
		return false, err
	}

	cpuTimes := jsonOrDefault(resultData["cpu_times"], map[string]any{})
	memory := jsonOrDefault(resultData["memory"], map[string]any{})
	diskIO := jsonOrDefault(resultData["disk_io"], []any{})

	if err = queries.UpsertHostRuntime(ctx, db.UpsertHostRuntimeParams{
		CreateTime: now, UpdateTime: now, HostID: hostID,
		CpuUsagePercent:       nullFloat64Of(floatOrNil(resultData["cpu_usage_percent"])),
		CpuTimes:              json.RawMessage(cpuTimes),
		MemoryUsagePercent:    nullFloat64Of(floatOrNil(resultData["memory_usage_percent"])),
		Memory:                json.RawMessage(memory),
		DiskIo:                json.RawMessage(diskIO),
		OsUptimeSeconds:       nullInt64Of(intOrNil(resultData["os_uptime_seconds"])),
		OsBootTime:            nullTimeOf(timeOrNil(resultData["os_boot_time"])),
		MetricsSampleWindowMs: nullInt32Of(intOrNil(resultData["metrics_sample_window_ms"])),
		StaticFingerprint:     fingerprint,
		CollectedAt:           sql.NullTime{Time: now, Valid: true},
	}); err != nil {
		return false, err
	}

	if previousFingerprint != "" && previousFingerprint == fingerprint {
		// Static assets unchanged: dynamic runtime snapshot above is already fresh, skip the rest.
		return true, nil
	}

	osType := stringOrEmpty(resultData["os_type"])
	if osType == "" {
		osType = stringOrEmpty(resultData["os"])
	}
	if err = queries.UpsertHostSystem(ctx, db.UpsertHostSystemParams{
		CreateTime: now, UpdateTime: now, HostID: hostID,
		OsType:          nullableStrField(osType),
		OsVersion:       nullableStrField(stringOrEmpty(resultData["os_version"])),
		OsID:            nullableStrField(strings.ToLower(stringOrEmpty(resultData["os_id"]))),
		OsIDLike:        nullableStrField(strings.ToLower(stringOrEmpty(resultData["os_id_like"]))),
		OsVersionID:     nullableStrField(stringOrEmpty(resultData["os_version_id"])),
		KernelVersion:   nullableStrField(stringOrEmpty(resultData["kernel_version"])),
		Hostname:        nullableStrField(stringOrEmpty(resultData["hostname"])),
		AgentVersion:    nullableStrField(stringOrEmpty(resultData["agent_version"])),
		TimezoneName:    nullableStrField(stringOrEmpty(resultData["os_timezone"])),
		UtcOffset:       nullableStrField(stringOrEmpty(resultData["os_utc_offset"])),
		CollectorSource: sql.NullString{String: "agent", Valid: true},
		CollectedAt:     sql.NullTime{Time: now, Valid: true},
	}); err != nil {
		return false, err
	}

	diskTotalGB := sql.NullFloat64{}
	total := 0.0
	for _, disk := range disks {
		if size, ok := disk["size_gb"].(float64); ok {
			total += size
		}
	}
	if total > 0 {
		diskTotalGB = sql.NullFloat64{Float64: float64(int(total*10+0.5)) / 10, Valid: true}
	}

	if err = queries.UpsertHostHardware(ctx, db.UpsertHostHardwareParams{
		CreateTime: now, UpdateTime: now, HostID: hostID,
		CpuCores:     nullInt32Of(intOrNil(resultData["cpu_count"])),
		CpuModel:     nullableStrField(stringOrEmpty(resultData["cpu_model"])),
		MemoryGb:     nullFloat64Of(floatOrNil(resultData["memory_total_gb"])),
		DiskTotalGb:  diskTotalGB,
		Architecture: nullableStrField(stringOrEmpty(resultData["arch"])),
		CollectedAt:  sql.NullTime{Time: now, Valid: true},
	}); err != nil {
		return false, err
	}

	// Disk table has no unique key per device: rebuild fully to drop unmounted/removed partitions.
	tx, err := pool.BeginTx(ctx, nil)
	if err != nil {
		return false, err
	}
	txQueries := db.New(tx)
	if err = txQueries.DeleteHostDisks(ctx, hostID); err != nil {
		tx.Rollback()
		return false, err
	}
	for _, disk := range disks {
		device, _ := disk["device"].(string)
		if device == "" {
			continue
		}
		mountPoint, filesystem := sql.NullString{}, sql.NullString{}
		if v, ok := disk["mount_point"].(string); ok && v != "" {
			mountPoint = sql.NullString{String: v, Valid: true}
		}
		if v, ok := disk["filesystem"].(string); ok && v != "" {
			filesystem = sql.NullString{String: v, Valid: true}
		}
		sizeGB, usedGB := sql.NullFloat64{}, sql.NullFloat64{}
		if v, ok := disk["size_gb"].(float64); ok {
			sizeGB = sql.NullFloat64{Float64: v, Valid: true}
		}
		if v, ok := disk["used_gb"].(float64); ok {
			usedGB = sql.NullFloat64{Float64: v, Valid: true}
		}
		if err = txQueries.CreateHostDisk(ctx, db.CreateHostDiskParams{
			HostID: hostID, Device: device, MountPoint: mountPoint, SizeGb: sizeGB,
			UsedGb: usedGB, Filesystem: filesystem,
		}); err != nil {
			tx.Rollback()
			return false, err
		}
	}
	if err = tx.Commit(); err != nil {
		return false, err
	}
	return true, nil
}

// normalizeHostDisks filters out unnamed/squashfs entries and sorts deterministically, matching
// Django's _normalize_disks() so the static fingerprint is stable regardless of report order.
func normalizeHostDisks(raw any) []map[string]any {
	items, _ := raw.([]any)
	normalized := make([]map[string]any, 0, len(items))
	for _, rawItem := range items {
		item, ok := rawItem.(map[string]any)
		if !ok {
			continue
		}
		device := strings.TrimSpace(stringOrEmpty(item["device"]))
		if device == "" {
			continue
		}
		filesystem := strings.TrimSpace(stringOrEmpty(item["filesystem"]))
		if strings.EqualFold(filesystem, "squashfs") {
			continue
		}
		entry := map[string]any{"device": device, "mount_point": strings.TrimSpace(stringOrEmpty(item["mount_point"]))}
		if filesystem != "" {
			entry["filesystem"] = filesystem
		}
		if v := floatOrNil(item["size_gb"]); v != nil {
			entry["size_gb"] = v
		}
		if v := floatOrNil(item["used_gb"]); v != nil {
			entry["used_gb"] = v
		}
		normalized = append(normalized, entry)
	}
	sort.Slice(normalized, func(i, j int) bool {
		if normalized[i]["device"] != normalized[j]["device"] {
			return fmt.Sprint(normalized[i]["device"]) < fmt.Sprint(normalized[j]["device"])
		}
		return fmt.Sprint(normalized[i]["mount_point"]) < fmt.Sprint(normalized[j]["mount_point"])
	})
	return normalized
}

// buildStaticFingerprint hashes the fields that HostSystem/HostHardware/HostDisk care about, so
// unchanged static assets can skip those writes on every periodic refresh (see Django's
// _build_static_fingerprint). The hash only needs to be stable within this process/table, not
// cross-language identical to Django's.
func buildStaticFingerprint(resultData map[string]any, disks []map[string]any) string {
	osType := stringOrEmpty(resultData["os_type"])
	if osType == "" {
		osType = stringOrEmpty(resultData["os"])
	}
	payload := map[string]any{
		"os_type":         nullableStr(osType),
		"os_version":      nullableStr(stringOrEmpty(resultData["os_version"])),
		"os_id":           nullableStr(strings.ToLower(stringOrEmpty(resultData["os_id"]))),
		"os_id_like":      nullableStr(strings.ToLower(stringOrEmpty(resultData["os_id_like"]))),
		"os_version_id":   nullableStr(stringOrEmpty(resultData["os_version_id"])),
		"kernel_version":  nullableStr(stringOrEmpty(resultData["kernel_version"])),
		"hostname":        nullableStr(stringOrEmpty(resultData["hostname"])),
		"agent_version":   nullableStr(stringOrEmpty(resultData["agent_version"])),
		"cpu_count":       intOrNil(resultData["cpu_count"]),
		"cpu_model":       nullableStr(stringOrEmpty(resultData["cpu_model"])),
		"memory_total_gb": floatOrNil(resultData["memory_total_gb"]),
		"arch":            nullableStr(stringOrEmpty(resultData["arch"])),
		"os_timezone":     nullableStr(stringOrEmpty(resultData["os_timezone"])),
		"os_utc_offset":   nullableStr(stringOrEmpty(resultData["os_utc_offset"])),
		"disks":           disks,
	}
	// encoding/json marshals map keys in sorted order, giving a stable hash input.
	encoded, _ := json.Marshal(payload)
	digest := sha256.Sum256(encoded)
	return hex.EncodeToString(digest[:])
}

// nullableStrField 把"空串即 NULL"的字段钉成 sqlc 的 sql.NullString
// （与 nullableStr 同语义，只是返回具体类型给参数结构体用）。
func nullableStrField(value string) sql.NullString {
	if value == "" {
		return sql.NullString{}
	}
	return sql.NullString{String: value, Valid: true}
}

// nullInt32Of / nullInt64Of / nullFloat64Of / nullTimeOf 把 resultData 里的 any 形态值
// （agent 上报的 JSON）转成 sqlc 需要的具体可空类型，nil 一律落 NULL。
func nullInt32Of(value any) sql.NullInt32 {
	switch typed := value.(type) {
	case int64:
		return sql.NullInt32{Int32: int32(typed), Valid: true}
	case int:
		return sql.NullInt32{Int32: int32(typed), Valid: true}
	case float64:
		return sql.NullInt32{Int32: int32(typed), Valid: true}
	default:
		return sql.NullInt32{}
	}
}

func nullInt64Of(value any) sql.NullInt64 {
	switch typed := value.(type) {
	case int64:
		return sql.NullInt64{Int64: typed, Valid: true}
	case int:
		return sql.NullInt64{Int64: int64(typed), Valid: true}
	case float64:
		return sql.NullInt64{Int64: int64(typed), Valid: true}
	default:
		return sql.NullInt64{}
	}
}

func nullFloat64Of(value any) sql.NullFloat64 {
	if typed, ok := value.(float64); ok {
		return sql.NullFloat64{Float64: typed, Valid: true}
	}
	return sql.NullFloat64{}
}

func nullTimeOf(value any) sql.NullTime {
	if typed, ok := value.(time.Time); ok {
		return sql.NullTime{Time: typed, Valid: true}
	}
	return sql.NullTime{}
}

func stringOrEmpty(value any) string {
	s, _ := value.(string)
	return strings.TrimSpace(s)
}

func nullableStr(value string) any {
	if value == "" {
		return nil
	}
	return value
}

func floatOrNil(value any) any {
	switch v := value.(type) {
	case float64:
		return v
	case string:
		parsed, err := strconv.ParseFloat(v, 64)
		if err != nil {
			return nil
		}
		return parsed
	default:
		return nil
	}
}

func intOrNil(value any) any {
	switch v := value.(type) {
	case float64:
		return int64(v)
	case string:
		parsed, err := strconv.ParseInt(v, 10, 64)
		if err != nil {
			return nil
		}
		return parsed
	default:
		return nil
	}
}

func timeOrNil(value any) any {
	text, ok := value.(string)
	if !ok || strings.TrimSpace(text) == "" {
		return nil
	}
	parsed, err := time.Parse(time.RFC3339Nano, text)
	if err != nil {
		return nil
	}
	return parsed.UTC()
}

func jsonOrDefault(value any, fallback any) []byte {
	if value == nil {
		encoded, _ := json.Marshal(fallback)
		return encoded
	}
	encoded, err := json.Marshal(value)
	if err != nil {
		encoded, _ = json.Marshal(fallback)
	}
	return encoded
}
